src/shims/fs.rs

```diff
@@ -261,6 +261,98 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.try_unwrap_io_result(result)
     }
 
+    fn stat(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+        buf_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if this.tcx.sess.target.target.target_os.to_lowercase() != "macos" {
+            throw_unsup_format!("The `stat` shim is only only available in the `macos` platform.")
+        }
+
+        let path_scalar = this.read_scalar(path_op)?.not_undef()?;
+        let path = this.read_os_str_from_c_str(path_scalar)?;
+
+        let buf = this.deref_operand(buf_op)?;
+
+        let metadata = match std::fs::metadata(path) {
+            Ok(metadata) => metadata,
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                return Ok(-1);
+            }
+        };
+
+        let file_type = metadata.file_type();
+
+        let mode_name = if file_type.is_file() {
+            "S_IFREG"
+        } else if file_type.is_dir() {
+            "S_IFDIR"
+        } else {
+            "S_IFLNK"
+        };
+
+        // FIXME: use Scalar::to_u16
+        let mode = this.eval_libc(mode_name)?.to_bits(Size::from_bits(16))? as u16;
+
+        let size = metadata.len();
+
+        let (access_sec, access_nsec) = extract_sec_and_nsec(metadata.accessed(), &mut 0, 0)?;
+        let (created_sec, created_nsec) = extract_sec_and_nsec(metadata.created(), &mut 0, 0)?;
+        let (modified_sec, modified_nsec) = extract_sec_and_nsec(metadata.modified(), &mut 0, 0)?;
+
+        let dev_t_layout = this.libc_ty_layout("dev_t")?;
+        let mode_t_layout = this.libc_ty_layout("mode_t")?;
+        let nlink_t_layout = this.libc_ty_layout("nlink_t")?;
+        let ino_t_layout = this.libc_ty_layout("ino_t")?;
+        let uid_t_layout = this.libc_ty_layout("uid_t")?;
+        let gid_t_layout = this.libc_ty_layout("gid_t")?;
+        let time_t_layout = this.libc_ty_layout("time_t")?;
+        let long_layout = this.libc_ty_layout("c_long")?;
+        let off_t_layout = this.libc_ty_layout("off_t")?;
+        let blkcnt_t_layout = this.libc_ty_layout("blkcnt_t")?;
+        let blksize_t_layout = this.libc_ty_layout("blksize_t")?;
+        let uint32_t_layout = this.libc_ty_layout("uint32_t")?;
+
+        // We need to add 32 bits of padding after `st_rdev` if we are in a 64-bit platform. To do
+        // this, we store `st_rdev` as a `c_long` instead of a `dev_t`.
```

Comment:
Uh, wait, you are using the wrong type for `st_rdev` then? I don't think we should do that. Padding should be filled with 0, not by extening some other field in size.

Reply:
but then I would have to write the array of immediates twice, or create a `Vec`.

Reply:
Can't you add a field that has size 0 on 32bit and size 4 (bytes) on 64bit?

Reply:
I'm trying